### ISSUE TITLE
Feature Proposal: add chktex support?

### ISSUE THREAD

**Issue by jabooth** [James-Yu/LaTeX-Workshop#39]
Not sure if this is best handled separately but since I've done a little work on this figured I would at least document it here in case it's of use.

In case you aren't familiar, [Chktex](http://www.nongnu.org/chktex/) is a LaTeX linter - it performs static analysis on `.tex` files and reports numerous warnings for many times of common issues people run into when writing LaTeX.

It was simple enough to setup a task in vscode to run chktex:
```json
{
    "version": "0.1.0",
    "tasks": [
        {
            "taskName": "build",
            "command": "chktex",
            "isShellCommand": true,
            "args": ["-q", "-v0", "thesis.tex"],
            "showOutput": "never",
            "problemMatcher": {
                "owner": "chktex",
                "fileLocation": ["relative", "${workspaceRoot}"],
                "pattern": {
                    "regexp": "^(.*):(\\d+):(\\d+):(\\d+):(.*)$",
                    "file": 1,
                    "line": 2,
                    "column": 3,
                    "code": 4,
                    "message": 5
                },
                "severity": "warning"
            }
        }
    ]
}
```
This parses results that look like this:
```
fitting/supplementary.tex:592:5:44:Wrong length of dash may have been used.
```
and turns them into Problems:
![image](https://cloud.githubusercontent.com/assets/1312873/24331426/1757e6f6-122c-11e7-90ec-b224226a1242.png)

I then used [triggertaskonsave](https://marketplace.visualstudio.com/items?itemName=Gruntfuggly.triggertaskonsave) to launch this every time I save a .tex file:
```json
    "triggerTaskOnSave.tasks": {
        "build": ["**/*.tex"]
    },
```
This setup actually works great and I'm totally happy with it. Saying that it's a little complex to put together for a lot of users, so wrapping this up in an extension would make it a lot more accessable, and make it portable across projects.

I considered making a Chktex extension to do this, but before I did that I wondered if you might be interested in a PR that adds Chktex support to LaTeX Workshop. I'd have to write a lot of the same boilerplate code to find the root tex file for instance, so I could see it making more sense as feature within LaTeX Workshop rather than it's own extension.

Ideally I think chktex linting would be an optional feature, that defaults to being on if they user has chktex installed (can always be disabled/cusomised in settings).

If enabled, chktex would be invoked for the root file on every save in parallel with the build, with results being parsed and shown in the problems window.

chktex is pretty fast, so it may even make sense to run chktex on the current file every time the user takes a short break from typing...again, I would suggest these settings would be customisable.

I took a quick look and I think I could quite easily add the feature by generalising the nice `Builder` class you have to maybe have a `Linter` variant.

Anyway, before getting into any more of the nitty gritty, how would you feel about such an addition?

**Comment by James-Yu** [comment]
I've never used that, yet it seems very promising. Will check how to integrate the new diagnostics with the original ones.

What I think now is that this chktex is called whenever user stopped typing for `configurable` seconds, and the generated log items are merged to the builder's. It's a bit late now at my region so will make commits on it tomorrow. Hope a `1.1.0` version with this feature as well as a `clean project` one.

**Comment by James-Yu** [comment]
Btw I'm not very satisfied on the current way of handling/storing diagnostics. Will work on it soon.

**Comment by jabooth** [comment]
Cool! Let me know how you get on, I'm more than happy to help with it!

**Comment by James-Yu** [comment]
I think this should be what you need. Mind have a test? If things works well, I will publish it on the marketplace.

**Comment by James-Yu** [comment]
just released as `1.1.0`. Try it out!

**Comment by jabooth** [comment]
@James-Yu great, testing it now! Found a few other issues, will raise issues/PRs.
